plutus-tx/src/Language/PlutusTx/Ratio.hs

```diff
@@ -0,0 +1,205 @@
+    (x :% y) <= (x' :% y') = x P.* y' P.<= (x' P.* y)
+
+{-# INLINABLE (%) #-}
+-- | Forms the ratio of two integral numbers.
+(%) :: Integer -> Integer -> Ratio Integer
+x % y = reduce (x P.* signum y) (abs y)
+
+-- | Convert a 'Data.Ratio.Rational' to a
+--   Plutus-compatible 'Language.PlutusTx.Ratio.Rational'
+fromGHC :: Ratio.Rational -> Ratio Integer
+fromGHC (n Ratio.:% d) = n :% d
+
+-- | Convert a 'Language.PlutusTx.Ratio.Rational' to a
+--   'Data.Ratio.Rational'
+toGHC :: Rational -> Ratio.Rational
+toGHC (n :% d) = n Ratio.:% d
+
+{-# INLINABLE numerator #-}
+-- | Extract the numerator of the ratio in reduced form: the numerator and denominator have no common factor and the denominator is positive.
+numerator :: Ratio a -> a
+numerator (n :% _) = n
+
+{-# INLINABLE denominator #-}
+-- | Extract the denominator of the ratio in reduced form: the numerator and denominator have no common factor and the denominator is positive.
+denominator :: Ratio a -> a
+denominator (_ :% d) = d
+
+{-# INLINABLE gcd #-}
+-- From GHC.Real
+-- | @'gcd' x y@ is the non-negative factor of both @x@ and @y@ of which
+-- every common factor of @x@ and @y@ is also a factor; for example
+-- @'gcd' 4 2 = 2@, @'gcd' (-4) 6 = 2@, @'gcd' 0 4@ = @4@. @'gcd' 0 0@ = @0@.
+gcd :: Integer -> Integer -> Integer
+gcd a b = gcd' (abs a) (abs b) where
+    gcd' a' b'
+        | b' P.== P.zero = a'
+        | True           = gcd' b' (a' `Builtins.remainderInteger` b')
+
+{-# INLINABLE truncate #-}
+-- | truncate @x@ returns the integer nearest @x@ between zero and @x@
+truncate :: Ratio Integer -> Integer
+truncate r = let (m, _ :: Rational) = properFraction r in m
+
+{-# INLINABLE properFraction #-}
+-- From GHC.Real
+-- | The function 'properFraction' takes a real fractional number @x@
+-- and returns a pair @(n,f)@ such that @x = n+f@, and:
+--
+-- * @n@ is an integral number with the same sign as @x@; and
+--
+-- * @f@ is a fraction with the same type and sign as @x@,
+--   and with absolute value less than @1@.
+--
+-- The default definitions of the 'ceiling', 'floor', 'truncate'
+-- and 'round' functions are in terms of 'properFraction'.
+properFraction :: Ratio Integer -> (Integer, Ratio Integer)
+properFraction (n :% d) = (q, r :% d) where (q, r) = quotRem n d
+
+{-# INLINABLE quot #-}
+quot :: Integer -> Integer -> Integer
+quot x y = let (q, _ :: Integer) = quotRem x y in q
+
+{-# INLINABLE quotRem #-}
+-- | simultaneous quot and rem
+quotRem :: Integer -> Integer -> (Integer, Integer)
+quotRem x y = 
+    -- `quot` is integer division truncated towards 0, while the result of `div`
+    -- is truncated toward negative inf.
+    -- Cf. https://stackoverflow.com/questions/339719/when-is-the-difference-between-quotrem-and-divmod-useful
+    -- So we need to add 1 in some cases!
+    let d   = x `Builtins.divideInteger` y
+        rem = x `Builtins.remainderInteger` y
+    in if d P.< P.zero P.&& rem P./= P.zero
+       then (d P.+ P.one, rem)
+       else (d, rem)
```

Comment:
I think is wrong (but probably not what's causing your problem).  I see that `divideInteger` is implemented as `div` and `remainderInteger` as `rem`.  You should never mix these two: either use `div` and `mod` or `quot` and `rem. I made a song and dance about this in the PLC spec but eventually took it out because it was annoying people.  If you have an integer division operation `//` and a remainder operation `%%` then you always want `n * (n//d) + (n %% d) = n` (d != 0), and this fails for `div` and `rem`.  For example, `(-41) div` 5  = -9` and `(-41) rem 5 = -1`, but `5*(-9)+(-1) = -46`.  Using either div/mod or quot/rem fixes this (but I still think they're both mathematically wrong!).

This stuff's really annoying and you have to be very careful.  I'd suggest having tests for all four possibilities for the  signs of numerator and denominator.

Reply:
Yeah it's odd that we have `div` and `rem`, but that's why I implemented `quotRem` here to express `quot` in terms of `div`. It would be much nicer if we had a `quotRem` builtin. 

All the code was copied from `GHC.Real` (see for exampe the definition of [`reduce`](https://hackage.haskell.org/package/base-4.12.0.0/docs/src/GHC.Real.html#reduce)) because I wanted the Plutus `Rational` to behave exactly like the Haskell `Rational`. That's what the test suite checks too.

Reply:
I realised that it was all the same as GHC.Real after a while, but it still worries me a bit.  I think that what was alarming me is that a lot of the code depends on the invariant that denominators are positive, and that isn't enforced by the representation, or indeed mentioned anywhere at all.  That seems a bit dangerous from a maintenance point of view.  If someone were to export `:%` for example then things would suddenly go badly wrong.

Also (but not due to this particular code), I think that only having `div` and `rem` is just plain wrong.  We should either have two consistent operations or all four, although having all four just leads to confusion.  As you point out, providing `quoteRem` or something would ensure that people don't mix incompatible operations.

Reply:
I added a note explaining that the code is really from `GHC.Real`, and I hid the constructor of `Ratio`.

Reply:
Thanks.  I might open an issue to replace `rem` with `mod` in `Builtins`.  That would make things a bit easier here.

Reply:
I don't think there's any particular reason to be attached to our current set of integer builtins.